File: app/src/main/java/com/cybersec/smartroute/engine/NativeRouteSimulator.java
package com.cybersec.smartroute.engine;

import com.cybersec.smartroute.model.LatLng;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Pure-Java route interpolator. Operates on a JSON session blob so it can
 * be driven by either {@link com.cybersec.smartroute.service.SpoofController}
 * (foreground) or the foreground service (background) using identical state.
 */
public final class NativeRouteSimulator {

    private final JSONObject session;
    private final String mode;
    private final List<LatLng> waypoints;
    private final LatLng start;
    private final LatLng end;
    private final boolean curved;
    private final double minSpeedKmh;
    private final double maxSpeedKmh;
    private final boolean enablePauses;
    private final double pauseProbability;
    private final String acceleration;

    private double progress;
    private double currentSpeedKmh;
    private int pauseTicksLeft;
    private boolean inPause;
    private final Random rng = new Random();

    public NativeRouteSimulator(JSONObject session) {
        this.session = session;
        this.mode = session.optString("mode", "dynamicPath");
        this.start = new LatLng(session.optDouble("startLat", 0), session.optDouble("startLon", 0));
        this.end = new LatLng(session.optDouble("endLat", 0), session.optDouble("endLon", 0));
        this.curved = session.optBoolean("curved", false);
        this.minSpeedKmh = session.optDouble("minSpeedKmh", 40);
        this.maxSpeedKmh = session.optDouble("maxSpeedKmh", 80);
        this.enablePauses = session.optBoolean("enablePauses", false);
        this.pauseProbability = session.optDouble("pauseProbability", 0.02);
        this.acceleration = session.optString("acceleration", "smooth");
        this.progress = session.optDouble("progress", 0);
        this.currentSpeedKmh = session.optDouble("currentSpeedKmh", minSpeedKmh);
        this.waypoints = parseWaypoints(session);
    }

    private static List<LatLng> parseWaypoints(JSONObject session) {
        List<LatLng> list = new ArrayList<>();
        JSONArray arr = session.optJSONArray("waypoints");
        if (arr != null) {
            for (int i = 0; i < arr.length(); i++) {
                JSONObject o = arr.optJSONObject(i);
                if (o != null) list.add(LatLng.fromJson(o));
            }
        }
        return list;
    }

    public boolean isComplete() {
        return progress >= 1.0;
    }

    /** Returns [lat, lon, done]. */
    public TickResult tick(double deltaSeconds) {
        if ("staticLocation".equals(mode)) {
            return new TickResult(start.latitude, start.longitude, false);
        }
        if (isComplete()) {
            return new TickResult(end.latitude, end.longitude, true);
        }
        if (pauseTicksLeft > 0) {
            pauseTicksLeft--;
            LatLng cur = positionAt(progress);
            return new TickResult(cur.latitude, cur.longitude, false);
        }
        if (enablePauses && !inPause && rng.nextDouble() < pauseProbability) {
            inPause = true;
            pauseTicksLeft = 2 + rng.nextInt(4);
            LatLng cur = positionAt(progress);
            return new TickResult(cur.latitude, cur.longitude, false);
        }
        inPause = false;

        double totalKm = totalRouteKm();
        if (totalKm <= 0) {
            progress = 1.0;
            return new TickResult(end.latitude, end.longitude, true);
        }

        double targetSpeed = targetSpeedKmh();
        currentSpeedKmh = applyAcceleration(currentSpeedKmh, targetSpeed, deltaSeconds);

        double distanceKm = (currentSpeedKmh / 3600.0) * deltaSeconds;
        progress = Math.max(0, Math.min(1, progress + distanceKm / totalKm));

        LatLng pos = positionAt(progress);
        return new TickResult(pos.latitude, pos.longitude, progress >= 1.0);
    }

    public JSONObject toJson() {
        JSONObject copy = new JSONObject();
        try {
            // Shallow clone via re-serialisation is sufficient since the
            // input object only carries primitives + an array.
            copy = new JSONObject(session.toString());
            copy.put("progress", progress);
            copy.put("currentSpeedKmh", currentSpeedKmh);
        } catch (JSONException ignored) {
        }
        return copy;
    }

    // ---------- helpers ----------

    private double totalRouteKm() {
        List<LatLng> pts = new ArrayList<>(waypoints.size() + 2);
        pts.add(start);
        pts.addAll(waypoints);
        pts.add(end);
        double total = 0;
        for (int i = 0; i < pts.size() - 1; i++) {
            total += LatLng.haversineKm(pts.get(i), pts.get(i + 1));
        }
        return total;
    }

    private LatLng positionAt(double globalT) {
        List<LatLng> pts = new ArrayList<>(waypoints.size() + 2);
        pts.add(start);
        pts.addAll(waypoints);
        pts.add(end);
        if (pts.size() < 2) return start;
        double[] lengths = new double[pts.size() - 1];
        double total = 0;
        for (int i = 0; i < lengths.length; i++) {
            lengths[i] = LatLng.haversineKm(pts.get(i), pts.get(i + 1));
            total += lengths[i];
        }
        if (total == 0) return start;
        double remaining = Math.max(0, Math.min(1, globalT)) * total;
        for (int i = 0; i < lengths.length; i++) {
            double len = lengths[i];
            if (remaining <= len || i == lengths.length - 1) {
                double localT = len == 0 ? 1.0 : Math.max(0, Math.min(1, remaining / len));
                LatLng a = pts.get(i);
                LatLng b = pts.get(i + 1);
                return curved ? LatLng.curved(a, b, localT, 0.15) : LatLng.linear(a, b, localT);
            }
            remaining -= len;
        }
        return end;
    }

    private double targetSpeedKmh() {
        double mid = 0.5;
        double edge = Math.abs(progress - mid) * 2;
        double lerp = minSpeedKmh + (maxSpeedKmh - minSpeedKmh) * (1 - edge * 0.3);
        return Math.max(minSpeedKmh, Math.min(maxSpeedKmh, lerp));
    }

    private double applyAcceleration(double current, double target, double dt) {
        switch (acceleration) {
            case "constant":
                return target;
            case "sudden":
                double sign = target == current ? 0 : (target > current ? 1 : -1);
                return Math.max(minSpeedKmh, Math.min(maxSpeedKmh, current + sign * 25 * dt));
            case "smooth":
            default:
                double rate = 8.0;
                return current + (target - current) * (1 - Math.exp(-rate * dt));
        }
    }

    public static final class TickResult {
        public final double lat;
        public final double lon;
        public final boolean done;

        public TickResult(double lat, double lon, boolean done) {
            this.lat = lat;
            this.lon = lon;
            this.done = done;
        }
    }
}

File: android/app/src/main/kotlin/com/cybersec/smartroute/NativeRouteSimulator.kt
package com.cybersec.smartroute

import org.json.JSONObject
import kotlin.math.*

/**
 * Native route interpolation — keeps mock GPS running if the Flutter isolate pauses.
 */
class NativeRouteSimulator(private val session: JSONObject) {
    private var progress = session.optDouble("progress", 0.0)
    private var currentSpeedKmh = session.optDouble("minSpeedKmh", 40.0)

    private val mode: String = session.optString("mode", "dynamicPath")
    private val startLat = session.getDouble("startLat")
    private val startLon = session.getDouble("startLon")
    private val endLat = session.getDouble("endLat")
    private val endLon = session.getDouble("endLon")
    private val safeLat = session.optDouble("safeLat", startLat)
    private val safeLon = session.optDouble("safeLon", startLon)
    private val minSpeed = session.optDouble("minSpeedKmh", 40.0)
    private val maxSpeed = session.optDouble("maxSpeedKmh", 80.0)
    private val curved = session.optBoolean("curved", false)

    val isComplete: Boolean get() = progress >= 1.0

    fun tick(deltaSeconds: Double): Triple<Double, Double, Boolean> {
        if (mode == "staticLocation") {
            return Triple(startLat, startLon, false)
        }
        if (isComplete) return Triple(endLat, endLon, true)

        val totalKm = haversineKm(startLat, startLon, endLat, endLon)
        if (totalKm <= 0.0) {
            progress = 1.0
            return Triple(endLat, endLon, true)
        }

        val target = minSpeed + (maxSpeed - minSpeed) * 0.5
        currentSpeedKmh += (target - currentSpeedKmh) * (1 - exp(-8.0 * deltaSeconds))

        val distanceKm = (currentSpeedKmh / 3600.0) * deltaSeconds
        progress = (progress + distanceKm / totalKm).coerceIn(0.0, 1.0)

        val t = progress
        val (lat, lon) = if (curved) {
            bezier(startLat, startLon, endLat, endLon, t)
        } else {
            linear(startLat, startLon, endLat, endLon, t)
        }
        return Triple(lat, lon, progress >= 1.0)
    }

    fun safeZone(): Pair<Double, Double> = safeLat to safeLon

    fun toJson(): JSONObject {
        val copy = JSONObject(session.toString())
        copy.put("progress", progress)
        return copy
    }

    private fun linear(
        sLat: Double, sLon: Double, eLat: Double, eLon: Double, t: Double,
    ): Pair<Double, Double> {
        val u = t.coerceIn(0.0, 1.0)
        return (sLat + (eLat - sLat) * u) to (sLon + (eLon - sLon) * u)
    }

    private fun bezier(
        sLat: Double, sLon: Double, eLat: Double, eLon: Double, t: Double,
    ): Pair<Double, Double> {
        val u = t.coerceIn(0.0, 1.0)
        val midLat = (sLat + eLat) / 2
        val midLon = (sLon + eLon) / 2
        val dy = eLat - sLat
        val dx = eLon - sLon
        val cLat = midLat + dy * 0.15
        val cLon = midLon - dx * 0.15
        val lat = (1 - u) * (1 - u) * sLat + 2 * (1 - u) * u * cLat + u * u * eLat
        val lon = (1 - u) * (1 - u) * sLon + 2 * (1 - u) * u * cLon + u * u * eLon
        return lat to lon
    }

    private fun haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6371.0
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2).pow(2) +
            cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(dLon / 2).pow(2)
        return 2 * r * asin(sqrt(a))
    }
}
